ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouterMetrics.java

```diff
@@ -31,42 +37,141 @@
+    invalidBlobIdErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "InvalidBlobIdErrorCount"));
+    invalidPutArgumentErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "InvalidPutArgumentErrorCount"));
+    operationTimedOutErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "OperationTimedOutErrorCount"));
+    routerClosedErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "RouterClosedErrorCount"));
+    unexpectedInternalErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "UnexpectedInternalErrorCount"));
+    blobTooLargeErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "BlobTooLargeErrorCount"));
+    badInputChannelErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "BadInputChannelErrorCount"));
+    insufficientCapacityErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "InsufficientCapacityErrorCount"));
+    blobDeletedErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "BlobDeletedErrorCount"));
+    blobDoesNotExistErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "BlobDoesNotExistErrorCount"));
+    blobExpiredErrorCount =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "BlobExpiredErrorCount"));
+    unknownReplicaResponseError =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "UnknownReplicaResponseError"));
+    closeErrorCount = metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "CloseErrorCount"));
+    unknownErrorCountForOperation =
+        metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "UnknownErrorCountForOperation"));
+
+    // Misc metrics.
+    operationErrorRate =
+        metricRegistry.meter(MetricRegistry.name(NonBlockingRouter.class, "OperationErrorArrivalRate"));
+    slippedPutSuccessCount = metricRegistry.counter(MetricRegistry.name(PutManager.class, "SlippedPutSuccessCount"));
+    ignoredResponseCount = metricRegistry.counter(MetricRegistry.name(NonBlockingRouter.class, "IgnoredRequestCount"));
+
+    // Track metrics at the DataNode level.
+    dataNodeToMetrics = new HashMap<DataNodeId, NodeLevelMetrics>();
+    for (DataNodeId dataNodeId : clusterMap.getDataNodeIds()) {
+      String dataNodeName = dataNodeId.getDatacenterName() + "." + dataNodeId.getHostname() + "." + Integer
+          .toString(dataNodeId.getPort());
+      dataNodeToMetrics.put(dataNodeId, new NodeLevelMetrics(metricRegistry, dataNodeName));
+    }
+    // A null key corresponds to the NodeLevelMetrics for all unknown DataNodeId.
+    dataNodeToMetrics.put(null, new NodeLevelMetrics(metricRegistry, "UnknownDataNode.UnknownHostName.UnknownPort"));
```

Comment:
there are no unknown datanodes. only those that are known (in the cluster map) are datanodes.